tool/planet/etcd.go

```diff
@@ -196,9 +200,97 @@ func etcdEnable(upgradeService bool) error {
 		log.Info("etcd is in proxy mode, nothing to do")
 		return nil
 	}
+
 	return trace.Wrap(enableService(ctx, ETCDUpgradeServiceName))
 }
 
+// etcdInitJoin ensures this particular node is part of an etcd cluster.
+// Because the etcd cluster is re-created during an upgrade, this particular node may not be part of the new cluster.
+func etcdInitJoin(ctx context.Context, initMaster string) error {
+	env, err := box.ReadEnvironment(ContainerEnvironmentFile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// if this is a proxy node, don't join the cluster
+	if env.Get(EnvEtcdProxy) == "on" {
+		return nil
+	}
+
+	// if this is a new cluster, we don't need to bootstrap membership, etcd will take care of this for us
+	if env.Get(EnvEtcdInitialClusterState) != "existing" {
+		return nil
+	}
+
+	// Gravity will give us a hint about which master to join if we need to run init
+	if initMaster == "" {
+		return nil
+	}
+
+	log.WithField("master", initMaster).Info("Joining etcd to existing cluster")
+
+	return trace.Wrap(utils.Retry(ctx, math.MaxInt64, 1*time.Second, func() error {
+		conf := etcdconf.Config{
+			Endpoints: []string{initMaster},
+			KeyFile:   DefaultEtcdctlKeyFile,
+			CertFile:  DefaultEtcdctlCertFile,
+			CAFile:    DefaultEtcdctlCAFile,
+		}
+		client, err := conf.NewClientV3()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		publicIP := env.Get(EnvPublicIP)
+		if len(publicIP) == 0 {
+			return trace.BadParameter("public ip env variable is empty").AddField(EnvPublicIP, publicIP)
+		}
+
+		advertisePeerUrl := fmt.Sprintf("https://%v:2380", publicIP)
+
+		resp, err := client.MemberList(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		peerUrls := []string{}
+
+		isMember := false
+		for _, member := range resp.Members {
```

Comment:
Any way to factor out this logic into a method like `func isMember(client etcd.Client, advertiseIP string) bool`?